DOCS-762: Add custom API creation info (#1535)
viamrobotics/docs: docs/extend/modular-resources/create-type.md

---
title: "Define a New Resource Type"
linkTitle: "New API Type"
weight: 30
type: "docs"
tags: ["rdk", "extending viam", "modular resources", "API"]
description: "Define an API for a new type of resource."
no_list: true
---

You need to define a new {{% glossary_tooltip term_id="resource" text="resource" %}} *type* (and an API for that type) if:

- You have a {{% glossary_tooltip term_id="resource" text="resource" %}} that does not fit into any of the existing [component](/components/) or [service](/services/) types.

> - You have a resource that could fit into an existing type, for example, [base](/components/base/), but you want to define an API with different endpoints than the ones in the standard Viam [API](/program/apis/) for that type.

Reviewer: I think a more explicit example would be really helpful here

Author: ```suggestion
- You have a resource that could fit into an existing type, but you want to define an API with different endpoints than the ones in the standard Viam [API](/program/apis/) for that type.
For example, you have a [sensor](/components/sensor/) and you want to create a `Calibrate` endpoint.
```

{{% alert title="Tip" color="tip" %}}

> If you want to use most of an existing API but need just a few other functions, try using the [`DoCommand`](/program/apis/#docommand) endpoint to add custom functionality to an existing type.

Reviewer: worth mentioned extraParams as well, here?

> If you want to use most of an existing API but need just a few other functions, try using the [`DoCommand`](/program/apis/#docommand) endpoint to add custom functionality to an existing type.

Author: ```suggestion
If you want to use most of an existing API but need just a few other functions, try using the [`DoCommand`](/program/apis/#docommand) endpoint and [extra parameters](/program/use-extra-params/) to add custom functionality to an existing type.
```

{{% /alert %}}

## How to define your new resource API

Viam uses [protocol buffers](https://protobuf.dev/) for API definition.

To define a new type, you need to define the messages and methods of the new API in [protobuf](https://github.com/protocolbuffers/protobuf), generate all necessary [protobuf module files](https://buf.build/docs/generate/usage/), and then write code in Python or Go to implement the higher level server and client functions required.
The following steps guide you through this process in more detail.

1. Decide whether your type is a component or a service.

>   If it is hardware, it is a component.

Reviewer: ```suggestion
  If it is providing an interface to control hardware, it is a component.
```

Author: ```suggestion
  If provides an interface to control hardware, it is a component.
```

>   If it is a software package, it is a service.

Reviewer: ```suggestion
  If it is providing higher-level functionality, it is a service.
```

Since all of it is really software, and some services are still interacting with the underlying hardware (for example, motion or vision)

Author: ```suggestion
  If it provides higher-level functionality, it is a service.
```

1. Choose a name for your type.
  For example, "gizmo."
1. Create a directory or repository for your module.
  Within that, create a directory called <file>src</file>.
1. Define your new API:

    - Write the proto methods in a `<type name>.proto` file inside your <file>src</file> directory
      - [Example component](https://github.com/viamrobotics/viam-python-sdk/blob/main/examples/module/src/proto/gizmo.proto)
      - [Example service](https://github.com/viam-labs/speech/blob/main/src/proto/speech.proto)
    - And define those methods in a protobuf-supported language such as Python or Go in a file called `api.py` (or `api.go`)
      - [Example component in Python](https://github.com/viamrobotics/viam-python-sdk/blob/main/examples/module/src/gizmo/api.py)
      - [Example service in Python](https://github.com/viam-labs/speech/blob/main/src/speech/api.py)

> 1. In the root directory of your module, you need to generate the following boilerplate files.

Author: ```suggestion
1. In the root directory of your module, you need to generate some boilerplate files.
The following list includes the files you will typically need for most modules, though different files are required for some advanced use cases.
```
